library/src/main/java/com/mux/player/media/MuxDataSource.kt

```diff
@@ -98,4 +101,209 @@ class MuxDataSource private constructor(
     cacheWriter?.close()
     upstream?.close()
   }
+
+  private fun openAndInitRevalidating(dataSpec: DataSpec, readHandle: ReadHandle): Long {
+    val revalidateRequestHeaders = mutableMapOf<String, String>()
+    revalidateRequestHeaders.putAll(dataSpec.httpRequestHeaders)
+    revalidateRequestHeaders["If-None-Match"] = readHandle.fileRecord.etag
+
+    val revalidateSpec = dataSpec.buildUpon()
+      .setHttpMethod(DataSpec.HTTP_METHOD_GET) // deviate from spec to save the round-trip w/cdn
+      .setHttpRequestHeaders(revalidateRequestHeaders)
+      .build()
+
+    val upstreamBytes = openAndInitFromRemote(revalidateSpec, RevalidatingDataSource.Factory())
+    val upstream = this.upstream!! // set by initAndOpenUpstream
+
+    Log.d(TAG, "revalidation: HTTP ${upstream.responseCode}. $upstreamBytes available")
+    return if (upstream.responseCode != 304) {
+      // Entry wasn't valid anymore, but we did a GET so the body's ready to read and we're done
+      Log.d(TAG, "revalidation: Entry was NOT valid, getting from network")
+
+      // todo - we *could* delete the row here, but consider that stale items can be used if
```

Comment:
question(blocking): Shouldn't we minimally check for 200 or similar "full" responses and replace the cached segment with the freshly-retrieved one?